File: 프로그래머스/2/42860. 조이스틱/조이스틱.cpp
#include <bits/stdc++.h>
using namespace std;
int prv[25];
int n;
int mn = 1e9;

void dfs(int idx, deque<int>& dq)
{
    if (idx == dq.size())
    {
        int cur = 0, res = 0;
        deque<int> sdq = dq;
        for (int i = 0; i < dq.size(); i++)
        {
            if (prv[i] == 0)
            {
                res += (cur - sdq.back() + n) % n;
                cur = sdq.back(); sdq.pop_back();
            }
            else
            {
                res += (sdq.front() - cur + n) % n;
                cur = sdq.front(); sdq.pop_front();
            }
        }
        
        mn = min(mn, res);
        return;
    }
    
    for (int i = 0; i < 2; i++)
    {
        prv[idx] = i;
        dfs(idx + 1, dq);
    }
}

int solution(string s) 
{
    n = s.size();
    int ans = 0;
    deque<int> dq;
    for (int i = 0; i < n; i++)
    {
        if (s[i] != 'A')
        {
            if (i > 0) dq.push_back(i);
            int r1 = (s[i] - 'A');
            int r2 = ('A' - s[i] + 26) % 26;
            ans += min(r1, r2);
        }
    }
    
    dfs(0, dq);
    return ans + mn;
}

File: 프로그래머스/2/42860. 조이스틱/조이스틱.java
import java.util.*;

class Solution 
{    
    public int solution(String s) 
    {
        int n = s.length();
        ArrayList<Integer> v = new ArrayList<>();
        int ans = 0;
        
        for (int i = 0; i < n; i++)
        {
            ans += Math.min(s.charAt(i) - 'A', ('A' - s.charAt(i) + 26) % 26);
            if (i > 0 && s.charAt(i) != 'A') 
                v.add(i);
        }
        
        if (v.size() == 0)
            return ans;
        
        int res = Math.min(v.get(v.size() - 1), n - v.get(0));
        
        for (int i = 0; i < v.size(); i++)
        {
            int li = v.get(i), ri = v.get((i + 1) % v.size());
            int v1 = li * 2 + (n - ri);
            int v2 = (n - ri) * 2 + li;
            res = Math.min(res, Math.min(v1, v2));
        }
        
        return ans + res;
    }
}
